sympy/codegen/ast.py

```diff
@@ -399,3 +448,515 @@ def body(self):
+        ValueError: Maximum value for data type smaller than new value.
+        >>> v10 = 12345.67894
+        >>> float32.cast_check(v10)  # doctest: +ELLIPSIS
+        Traceback (most recent call last):
+          ...
+        ValueError: Casting gives a significantly different value.
+        >>> from sympy.codegen.ast import float64
+        >>> float64.cast_check(v10)
+        12345.67894
+        >>> from sympy import Float
+        >>> v18 = Float('0.123456789012345646')
+        >>> float64.cast_check(v18)
+        Traceback (most recent call last):
+          ...
+        ValueError: Casting gives a significantly different value.
+        >>> from sympy.codegen.ast import float80
+        >>> float80.cast_check(v18)
+        0.123456789012345649
+
+        """
+        from sympy.functions.elementary.complexes import im, re
+        val = _sympify(value)
+
+        ten = Integer(10)
+        exp10 = getattr(self, 'decimal_dig', None)
+
+        if rtol is None:
+            rtol = 1e-15 if exp10 is None else 2.0*ten**(-exp10)
+
+        def tol(num):
+            return atol + rtol*abs(num)
+
+        new_val = self._cast_nocheck(value)
+        self._check(new_val)
+
+        delta = new_val - val
+        if abs(delta) > tol(val):  # rounding, e.g. int(3.5) != 3.5
+            raise ValueError("Casting gives a significantly different value.")
+
+        return new_val
+
+
+class IntBaseType(Type):
+    """ Integer base type, contains no size information. """
+    __slots__ = ['name']
+    _cast_nocheck = Integer
+
+
+class _SizedIntType(IntBaseType):
+    __slots__ = ['name', 'bits']
+
+    def _check(self, value):
+        if value < self.min:
+            raise ValueError("Value is too small: %d < %d" % (value, self.min))
+        if value > self.max:
+            raise ValueError("Value is too big: %d > %d" % (value, self.max))
+
+
+class SignedIntType(_SizedIntType):
+    @property
+    def min(self):
+        return -2**(self.bits-1)
+
+    @property
+    def max(self):
+        return 2**(self.bits-1) - 1
+
+
+class UnsignedIntType(_SizedIntType):
+    @property
+    def min(self):
+        return 0
+
+    @property
+    def max(self):
+        return 2**self.bits - 1
+
+
+class FloatType(Type):
+    """ Represents a floating point value.
+
+    Examples
+    --------
+    >>> from sympy import S, Float
+    >>> from sympy.codegen.ast import FloatType
+    >>> half_precision = FloatType('f16', bits=16, max=65504, tiny=S(2)**-14,
+    ...     eps=Float('0.0009765625'), dig=2, decimal_dig=4)
+    >>> half_precision.cast_check(1.0)
+    1.0
+    >>> half_precision.cast_check(1e5)  # doctest: +ELLIPSIS
+    Traceback (most recent call last):
+      ...
+    ValueError: Maximum value for data type smaller than new value.
+
+    """
+    __slots__ = ['name', 'bits', 'max', 'tiny', 'eps', 'dig', 'decimal_dig']
+
+    def _cast_nocheck(self, value):
+        return Float(str(_sympify(value).evalf(self.decimal_dig)), self.decimal_dig)
```

Comment:
I was thinking that symbols, e.g. 'x', should be disallowed. But the user could of course pass in e.g. `'123.456'`, changing this to `sympify`